### PR TITLE
`--[lang]-query` can read tree-sitter (scm) query files

### PR DESCRIPTION
**Pull request by bheylin**
Relates to #143 

@alexpovel I've added a proof of concept to this PR.
The most logical place to add the file check and read is to the `struct CustomRustQuery`.
If so the `QueryError` needs to be wrapped in a compound `Error` that can report 
`io::Error`'s along with the parsing errors `QueryError` is now designed to report.

What do you think?

### CODE REVIEW

File: src/scoping/langs/rust.rs [hunk 1]
```diff
@@ -350,7 +352,17 @@ impl FromStr for CustomRustQuery {
     type Err = QueryError;
 
     fn from_str(s: &str) -> Result<Self, Self::Err> {
-        match TSQuery::new(&Rust::lang(), s) {
+        let path = Path::new(s);
+
+        let s: Cow<'_, str> = if path.is_file() {
+            fs::read_to_string(path)
+                .expect("How should QueryError be modified?")
```
Comment (alexpovel) [inline src/scoping/langs/rust.rs:359]: `QueryError` is part of tree-sitter, and cannot be modified here. I also don't think it should: `from_str` shouldn't be doing I/O. I'm trying to push I/O out as far as possible, aka keeping it out of `lib.rs` and children and contain it in `main.rs`. There, suitable error types (should) already exist, or could be added easily.

Comment (alexpovel) [reply src/scoping/langs/rust.rs:359]: https://github.com/tree-sitter/tree-sitter/blob/8500e331ebfd49e66dd935b8a9c7a58aba68af37/lib/binding_rust/lib.rs#L256

Comment (bheylin) [reply src/scoping/langs/rust.rs]: > `QueryError` is part of tree-sitter, and cannot be modified here.

Ah I didn't notice I had drifted into `tree-sitter` when I was looking at `QueryError`.

>  I also don't think it should: `from_str` shouldn't be doing I/O. I'm trying to push I/O out as far as possible, aka keeping it out of `lib.rs` and children and contain it in `main.rs`. There, suitable error types (should) already exist, or could be added easily.

I agree that `FromStr` and `From` impls should not be performing side-effects such as I/O but the current bounds on `CodeQuery` are make it quite difficult / impossible to move the I/O outside of this pipeline.

I've loosened up the pipeline in https://github.com/alexpovel/srgn/pull/144/commits/993c59dd083a550b257b795a6ccec8c1fbd4b4bd.

This involves making a `cli::CodeQuery` and a related `lang::CodeQuery` to keep a similar level of type safety but allow the pipeline to be intercepted in `main.rs` so I can try load a query file based on the path potentially contained in `cli::CodeQuery`. 

See what you think

File: src/scoping/langs/mod.rs [hunk 2]
```diff
@@ -48,77 +55,61 @@ pub struct Language<Q> {
...
     /// passed to Go etc.), so use this field to be more type-safe.
-    _marker: PhantomData<Q>,
+    _marker: PhantomData<T>,
 }
 
-impl<Q> Language<Q>
-where
-    Q: Into<TSQuery> + Clone,
-{
+impl<T: Kind> Language<T> {
```
Comment (alexpovel) [inline src/scoping/langs/mod.rs:61]: Could this be simpler? Paraphrasing (might not work!):

```rust
impl<T: Into<TSLanguage>> Language<T> {
```

where `TSLanguage` is:

https://github.com/tree-sitter/tree-sitter/blob/079c69313fa14b9263739b494a47efacc1c91cdc/lib/binding_rust/lib.rs#L61

and then we can just pass e.g. `tree_sitter_rust::LANGUAGE` into that, which implements it via:

https://github.com/tree-sitter/tree-sitter/blob/079c69313fa14b9263739b494a47efacc1c91cdc/lib/binding_rust/lib.rs#L423

then inside of `new` can call `T.into()` (might need turbofish?) instead of `&T::ts_lang()`.

Advantages I have in mind: saves the roundtrip through `LangKind` structs, and no `Kind` trait necessary.

But again, just a quick idea which might not work.

And note: sadly, HCL is vendored here as it's not published as a crate properly. I'm open to changing that! I can look into publishing a `srgn-tree-sitter-hcl` crate, so all languages can be treated the same (HCL is currently the odd one out and slightly different, which is annoying).

File: src/main.rs [hunk 3]
```diff
@@ -797,15 +808,30 @@ impl fmt::Display for ScoperBuildError {
 impl Error for ScoperBuildError {}
 
 #[allow(clippy::cognitive_complexity)] // 🤷‍♀️ macros
-fn get_language_scopers(args: &cli::Cli) -> Vec<Box<dyn LanguageScoper>> {
+fn get_language_scopers(args: &cli::Cli) -> Result<Vec<Box<dyn LanguageScoper>>, ProgramError> {
+    fn read_query_file(query_or_path: &cli::CodeQuery, buf: &mut String) -> io::Result<Option<()>> {
```
Comment (alexpovel) [inline src/main.rs:812]: Allocating inside of `read_query_file` and returning results (if found) should be totally fine. This only runs at startup, so performance is not an issue. This might clean up this functions usage (i.e. allows pattern matching over `is_some` further down IIUC)

File: src/main.rs [hunk 4]
```diff
@@ -1252,7 +1287,7 @@ mod cli {
 
         /// Scope C code using a custom tree-sitter query.
         #[arg(long, env, verbatim_doc_comment, value_name = TREE_SITTER_QUERY_VALUE_NAME)]
-        pub c_query: Vec<CustomCQuery>,
+        pub c_query: Vec<CodeQuery>,
```
Comment (alexpovel) [inline src/main.rs:1290]: So now `CodeQuery` is "less typesafe": all languages share a single type. How do we ensure we do not accidentally pass e.g. a Python tree-sitter query string to something expecting a different language? (that likely fails but MIGHT NOT, at which point confusion ensues)

**Review by alexpovel — COMMENTED** [review]
Nice work, thanks.

I was able to give it a first quick look, but not fully groked yet.

The `CodeQuery` `enum` was always bizarre, never liked it. It had this weird cyclic feel to it regarding trait impls. I get the feeling your refactor cleared that up.

It might be 2-3 weeks until I have time to look again properly, sorry! Feel free to reiterate, I will still take a look and unlock CI runs.

**Comment by alexpovel** [comment]
`Main / Execute release chores (pull_request)` is an expected failure, no worries there.

File: src/main.rs [hunk 5]
```diff
@@ -1361,6 +1396,25 @@ mod cli {
             <Self as CommandFactory>::command()
         }
     }
+
+    /// The given query arg could be a literal tree-sitter query or a path to a file containing
+    /// a tree-sitter query.
+    #[derive(Debug, Clone)]
+    pub struct CodeQuery(String);
+
+    impl Deref for CodeQuery {
```
[the remark this answers is not in the document]
Comment (bheylin) [reply src/main.rs]: Fixed in https://github.com/alexpovel/srgn/pull/144/commits/a9e1a7771b44b791c161da31d98dc7eea4aeb5e1

I'm not overly worried about the `Deref` anti-pattern myself but in this case it can easily be replaced with an `impl Into`.

File: src/main.rs (hunk 3 above)
Comment (bheylin) [reply src/main.rs]: Fixed in https://github.com/alexpovel/srgn/pull/144/commits/a9e1a7771b44b791c161da31d98dc7eea4aeb5e1

File: src/main.rs (hunk 4 above)
Comment (bheylin) [reply src/main.rs:1286]: The only place I see where an bug can occur is in the mapping of commands to languages in the `handle_language_scope` macro. And that mapping is simple to verify as correct by reading the mappings.

You can always add a `PhantomData<Lang>` to `CodeQuery` but for me it's not winning much. It feels like typing for the sake of typing.

Do you see some other issue that stronger typing would prevent?

File: src/scoping/langs/mod.rs (hunk 2 above)
Comment (bheylin) [reply src/scoping/langs/mod.rs:61]: Fair point. I'll look into an `<T: Into<TSLanguage>>` based solution.

Comment (bheylin) [reply src/scoping/langs/mod.rs]: I've made some quite large changes in https://github.com/alexpovel/srgn/pull/144/commits/d6fc6af3e724178a6f262ad8b27d98a7ca5c37ce.

I'll try and explain my changes below. Please see this commit as a proposal rather than me trying to demand changes.

I think that the `PhantomDatd` field in `scoping::langs::Language` is causing more trouble than it's worth.
The generic param `Q` can't be fulfilled by any of the `tree-sitter` language types as  they are not a struct unique to the given language. The type `tree-sitter::LanguageFn` can't be used as a generic param as it's not a struct or enum. So you end up needing to define a local struct per language mod to be used for the `Q` param.

I opted to just manually define a 'struct Language' per language mod instead of defining an alias of a `PhantomData` typed struct. So in short I replaced the generic `Language<Q>` with a per mod struct that composes a non-generic `Language`. 

This commit is made more complex by renaming the `struct Language` to a name I consider more representitive of how it's being used. I  propose renaming `struct Language<Q>` to `struct CompiledQuery`. 
Compiled query as opposed to a `RawQuery`.

This renaming makes the code simpler to grok for me at least.
See what you think yourself.

File: tests/langs/mod.rs [hunk 6]
```diff
@@ -73,763 +66,764 @@ impl InScopeLinePart {
...
     "base.c_callexpr",
     include_str!("c/base.c"),
-    C::new(CodeQuery::Prepared(PreparedCQuery::CallExpression)),
+    c::CompiledQuery::new(&c::PreparedQuery::CallExpression.into()),
 )]
 fn test_language_scopers(
     #[case] snapshot_name: &str,
     #[case] contents: &str,
-    #[case] lang: impl LanguageScoper,
+    #[case] lang: Result<impl LanguageScoper, tree_sitter::QueryError>,
```
Comment (alexpovel) [inline tests/langs/mod.rs:824]: nit: `tree_sitter::QueryError` -> `langs::TSQueryError`, aka the reexported one from `langs/mod.rs`?

File: src/main.rs [hunk 7]
```diff
@@ -651,6 +644,8 @@ enum ProgramError {
     IoError(io::Error),
     /// Error while processing files for walking.
     IgnoreError(ignore::Error),
+    /// The givon query failed to parse
```
Comment (alexpovel) [inline src/main.rs:647]: nit:

```suggestion
    /// The given query failed to parse.
```

File: src/main.rs (hunk 4 above)
Comment (alexpovel) [reply src/main.rs:1273]: Nope, makes sense

File: src/main.rs [hunk 8]
```diff
@@ -691,6 +689,12 @@ impl From<ignore::Error> for ProgramError {
     }
 }
 
+impl From<tree_sitter::QueryError> for ProgramError {
```
Comment (alexpovel) [inline src/main.rs:692]: Similar, can use `TSQueryError` (main motivation: "See all references" on `TSQueryError` then finds everything)

**Review by alexpovel — COMMENTED** [review]
The changes look good, thank you so much!

The `PhantomData` approach existed to be extra strict. I expected those types to be shuffled about a lot, where confusing them would be bad. But turns out that doesn't really happen, with construction only occurring in a single place.

Getting rid of the weirdly cyclic `CodeQuery` constructs is a breath of fresh air and appreciated. I'm happy to move forward with this PR if you are.

File: src/scoping/langs/mod.rs (hunk 2 above)
Comment (alexpovel) [reply src/scoping/langs/mod.rs]: I commented on this a bit already, but sounds good to me!

I had introduced the struct with `PhantomData` to be extra strict, and extra neat, but it never played well with the `tree_sitter` API (leaky abstraction), and ended up confusing, while not helping as you note.

File: src/main.rs (hunk 8 above)
Comment (bheylin) [reply src/main.rs:693]: Fixed in https://github.com/alexpovel/srgn/pull/144/commits/5c08ce08de67f30c77d691e4c0309efbe31f2bec

File: src/main.rs (hunk 7 above)
Comment (bheylin) [reply src/main.rs]: Fixed in https://github.com/alexpovel/srgn/pull/144/commits/5c08ce08de67f30c77d691e4c0309efbe31f2bec

File: tests/langs/mod.rs (hunk 6 above)
Comment (bheylin) [reply tests/langs/mod.rs]: Fixed in https://github.com/alexpovel/srgn/pull/144/commits/5c08ce08de67f30c77d691e4c0309efbe31f2bec

File: src/main.rs (hunk 7 above)
Comment (bheylin) [reply src/main.rs]: I'll make a new PR that adds a config file for the [`typos`](https://github.com/crate-ci/typos) CLI tool if you don't have any objections?

**Comment by bheylin** [comment]
> The changes look good, thank you so much!
> 

No worries, happy to help out.

> The `PhantomData` approach existed to be extra strict. I expected those types to be shuffled about a lot, where confusing them would be bad. But turns out that doesn't really happen, with construction only occurring in a single place.
> 
> Getting rid of the weirdly cyclic `CodeQuery` constructs is a breath of fresh air and appreciated. I'm happy to move forward with this PR if you are.

I'm happy to merge this into main when you are :wink:

File: src/main.rs (hunk 7 above)
Comment (alexpovel) [reply src/main.rs]: Yes feel free to go ahead! Love catching typos.

**Comment by alexpovel** [comment]
There are still some doc test failures in CI: https://github.com/alexpovel/srgn/actions/runs/11314113061/job/31463705601?pr=144#step:7:928

**Comment by bheylin** [comment]
> There are still some doc test failures in CI: https://github.com/alexpovel/srgn/actions/runs/11314113061/job/31463705601?pr=144#step:7:928

Fixed in https://github.com/alexpovel/srgn/pull/144/commits/2b2ea9d408b1e078c08cd25a7701ec7a1c665d5e

**Comment by bheylin** [comment]
@alexpovel I'll also add documentation for the new file mode of `--lang-query`.

**Comment by alexpovel** [comment]
Nice, thank you. One possible approach is:

- place a file with a language query somewhere in `docs/`, e.g, `docs/query`
- write Markdown with some dummy source code, with the `file=somefile.someextension` syntax you'll find for the other code blocks in the README, e.g. https://github.com/alexpovel/srgn/blob/9238fc161ceaf88dd754d71b4fb65393dcee964b/README.md#L123C9-L123C22
- invoke in a `console` Markdown code block as `srgn --some-lang-query docs/query --more-args`, then show results

This will all be picked up as part of unit testing via

https://github.com/alexpovel/srgn/blob/9238fc161ceaf88dd754d71b4fb65393dcee964b/tests/readme.rs

Lastly, docs like

https://github.com/alexpovel/srgn/blob/9238fc161ceaf88dd754d71b4fb65393dcee964b/src/main.rs#L1303

and

https://github.com/alexpovel/srgn/blob/9238fc161ceaf88dd754d71b4fb65393dcee964b/src/main.rs#L1224

(should read something like `TREE-SITTER-QUERY OR FILENAME` I guess) also need adjustment. That will trigger the README doc tests to fail, which you can then update (a manual process, sorry):

https://github.com/alexpovel/srgn/blob/9238fc161ceaf88dd754d71b4fb65393dcee964b/README.md#L1341

**Comment by alexpovel** [comment]
Oh and I'll squash all commits! If you feel like it you can do that yourself as well, otherwise I'll squash and come up with a suitable commit message.

Ideally and alternatively, instead of squashing into 1 commit, the languages refactor is one `chore` commit, and the addition of parsing queries from files is a second `feat` one, but it's fine if you don't feel like doing that work.

**Comment by bheylin** [comment]
> Oh and I'll squash all commits! If you feel like it you can do that yourself as well, otherwise I'll squash and come up with a suitable commit message.
> 
> Ideally and alternatively, instead of squashing into 1 commit, the languages refactor is one `chore` commit, and the addition of parsing queries from files is a second `feat` one, but it's fine if you don't feel like doing that work.

Yea no worries I'll squash the commits down to something reasonable.

**Comment by bheylin** [comment]
All good, I'll give this a try and maybe come back with questions

**Comment by bheylin** [comment]
@alexpovel as a start of making a test for the `--[lang]-query` flag, what do you think about moving the `[lang]::PreparedQuery` strings out to the `data/queries/[lang]/*.scm` location?

See https://github.com/alexpovel/srgn/pull/144/commits/94f68dc069fbebe5a11eae85c09688f8b7a1009c

This will make the query data easily available for various uses.

**Comment by alexpovel** [comment]
I am convinced 🙂 This guy will cause trouble though:

https://github.com/alexpovel/srgn/blob/9238fc161ceaf88dd754d71b4fb65393dcee964b/src/scoping/langs/mod.rs#L129

It is interpolated via `formatcp!` in places. I am fine with hard-coding the string in `.scm` files though. Would not be DRY as it is now (`_SRGN_IGNORE` is defined in a single place only) but easy to find and adjust.

**Comment by bheylin** [comment]
> I am convinced 🙂 This guy will cause trouble though:
> 
> https://github.com/alexpovel/srgn/blob/9238fc161ceaf88dd754d71b4fb65393dcee964b/src/scoping/langs/mod.rs#L129
> 
> It is interpolated via `formatcp!` in places. I am fine with hard-coding the string in `.scm` files though. Would not be DRY as it is now (`_SRGN_IGNORE` is defined in a single place only) but easy to find and adjust.

Yea there's no way around the use of `formatcp!` I'm afraid.
It's not necessary to move all the prepared strings out to `*.scm` files, I'm trying to avoid creating queries in the docs when we already have a lot of prepared queries, they're just bound up in the source code.

It may be worth it already though as, I think we could then modify the `rttest`s for each `[lang]::PreparedQuery` to read in the entire `data/queries/[lang]/` dir instead of enumerating each `[lang]::PreparedQuery` varinant:

```rust
#[rstest]
#[case(
    "base.py_comments",
    include_str!("python/base.py"),
    python::CompiledQuery::new(&python::PreparedQuery::Comments.into()),
)]
```

I don't have a lot of experience with `rstest`, but I think that's possible?

**Comment by bheylin** [comment]
And having the queries as files would allow the `--[lang]-query` sections in the `README.md` to be generated.

```markdown
      --rust <RUST>
          Scope Rust code using a prepared query.
          
          [env: RUST=]
          [aliases: rs]

          Possible values:
```

**Comment by bheylin** [comment]
@alexpovel We could move the prepared (`--rust`) and file based (`--rust-query`) tests into the `cli` tests mod.

It's moving to a more black-box test, but you can always choose do perform the tests on the lib API _and_ the CLI.

See: https://github.com/alexpovel/srgn/pull/144/commits/e24d015b0e5d48aeed5ac2cc0002e81a589192f8

**Comment by bheylin** [comment]
@alexpovel it's probably a good idea to keep testing the entire set of prepared queries in the lib itself so that the libs API is tested and for the CLI just test that a single prepared, inline and file based query works.

**Comment by alexpovel** [comment]
A couple thoughts:

- things like copy-pasting new `rstest` cases such as https://github.com/alexpovel/srgn/blob/3b8abfa38fe7ce94b3cdd446d76e85ecc4c2cf86/tests/langs/mod.rs#L72-L77 I am not worried about. That takes seconds to minutes for new features or languages. I understand your pain right now but it's not a common thing!
- similarly for e.g. generating the `srgn --help` output in the README. I've been adjusting that manually, which is also quick to do.
- as for exporting queries to `scm` files, I'm leaning towards not doing that. `_SRGN_IGNORE` is one example of where it starts to become a problem. The bigger issue I have is looking into the future, I want to remain dynamic, e.g. with string interpolation as it is now (`formatcp`), but also perhaps interpolation at runtime. I want to keep that option open. It would allow for a whole new level of functionality `srgn` can offer over base tree-sitter (queries). One example is pre- and post-processing for finer control.

    Long story short, "outsourced" `scm` files seem to stand in the way of that more than what they're helping
- testing prepared queries in the Rust library is a good point! That we can look into. On the other hand, I prefer true-as-possible end-to-end tests wherever possible. At the end of the day, `srgn` is a CLI first and we should guarantee correctness there as much as possible. The current test suite has very high coverage of all sorts of scenarios (snapshot tests make it ergonomic), and importantly is tested on all three platforms in CI. Moving more tests into just the library will take away from that coverage/thoroughness and decrease confidence (I've been quite happy and confident that when CI is green, `srgn` is good to go)

**Comment by bheylin** [comment]
> A couple thoughts:
> 
>     * things like copy-pasting new `rstest` cases such as https://github.com/alexpovel/srgn/blob/3b8abfa38fe7ce94b3cdd446d76e85ecc4c2cf86/tests/langs/mod.rs#L72-L77
>        I am not worried about. That takes seconds to minutes for new features or languages. I understand your pain right now but it's not a common thing!

Oh I wasn't worried about the effort it takes to add a new test, more that we could reduce the boilerplate.
But I take your point.

>     * similarly for e.g. generating the `srgn --help` output in the README. I've been adjusting that manually, which is also quick to do.
> 

All good, it's not a big issue.

>     * as for exporting queries to `scm` files, I'm leaning towards not doing that. `_SRGN_IGNORE` is one example of where it starts to become a problem. The bigger issue I have is looking into the future, I want to remain dynamic, e.g. with string interpolation as it is now (`formatcp`), but also perhaps interpolation at runtime. I want to keep that option open. It would allow for a whole new level of functionality `srgn` can offer over base tree-sitter (queries). One example is pre- and post-processing for finer control.
>       Long story short, "outsourced" `scm` files seem to stand in the way of that more than what they're helping
>

I'll roll back the last two commits and pursue another solution. 
 
>     * testing prepared queries in the Rust library is a good point! That we can look into. On the other hand, I prefer true-as-possible end-to-end tests wherever possible. At the end of the day, `srgn` is a CLI first and we should guarantee correctness there as much as possible. The current test suite has very high coverage of all sorts of scenarios (snapshot tests make it ergonomic), and importantly is tested on all three platforms in CI. Moving more tests into just the library will take away from that coverage/thoroughness and decrease confidence (I've been quite happy and confident that when CI is green, `srgn` is good to go)

It's good to know your attitude here, that the CLI tests are the most important.

**Comment by bheylin** [comment]
> Oh and I'll squash all commits! If you feel like it you can do that yourself as well, otherwise I'll squash and come up with a suitable commit message.
> 
> Ideally and alternatively, instead of squashing into 1 commit, the languages refactor is one `chore` commit, and the addition of parsing queries from files is a second `feat` one, but it's fine if you don't feel like doing that work.

@alexpovel I think the handiest is for me to make a new PR with the major renaming etc and rebase that onto this PR.
Then the history will be clean.

The working week has begun here so I'll try make some time for it during the week some evening.

**Comment by bheylin** [comment]
@alexpovel After merging #151 this is the resulting change needed to be able to provide a query as a file to `--[lang]-query`.

I'll add a test as discussed before and address comment https://github.com/alexpovel/srgn/pull/144#discussion_r1788675291.

**Comment by bheylin** [comment]
@alexpovel I've added a test to the `README.md`. I had the quote the file arg as the `readme.rs` parser does not support files with a leading dir (`srgn --python-query docs/python_cond_query.scm`). It took me a while to figure out what was going on there.

I'll update the docs for the CLI args now and call it a day.

How do you recommend testing the case where the file path doesn't exist etc?
`srgn --python-query docs/nonexistent-file.scm`.

File: src/main.rs [hunk 9]
```diff
@@ -1438,6 +1472,24 @@ mod cli {
         pub german_naive: bool,
     }
 
+    /// The query as read in from the CLI. This could either be a literal query or
+    /// a path to a query file.
+    #[derive(Clone, Debug)]
+    struct QuerySourceOrPath(String);
```
Comment (alexpovel) [inline src/main.rs:1478]: Just thinking out loud here... can something like

```rust
enum QuerySourceOrPath {
    Unknown(String),
    Query(RawQuery), // or `TSQuery`, or...
    Path(PathBuf),
}
```

make sense? (no need to bikeshed this though, it's fine as is)

**Review by alexpovel — COMMENTED** [review]
I'd be fine not testing the "non-existent file" scenario, as it's kinda "obviously correct" (bubbles the error up, as many other I/O errors do that can occur in `main.rs`, most of which aren't tested).

The error message for a non-existent file is a bit confusing of course:

```
Error: Error occurred while creating a tree-sitter query: Query error at 1:1. Invalid syntax:
docs/python_cond_query-file-does-not-exist.scm
^
```

but...

- if you're already passing a file name, you must have read the docs which mention that this is even possible. So you know what you're doing at will quickly be able to tell that the file doesn't exist
- I don't know how to report this better. On "file not found" error, we could check "does this LOOK like a file path", but I think that's fraught with peril... maybe a very simple, best-effort regex roughly like `/?([a-zA-Z0-9]+/)+`, and then if that matches `warn!()` the user.

  Alternatively, is it possible the queries ALWAYS have to start with `(`? If yes, that could also be a give-away.

File: src/main.rs [hunk 10]
```diff
@@ -1363,8 +1392,9 @@ mod cli {
 
         #[allow(clippy::doc_markdown)] // CamelCase detected as 'needs backticks'
         /// Scope HashiCorp Configuration Language code using a custom tree-sitter query.
-        #[arg(long, env, verbatim_doc_comment, value_name = TREE_SITTER_QUERY_VALUE_NAME)]
-        hcl_query: Vec<RawQuery>,
+        /// The query can be given inline or if as a path to a file containing a query.
```
Comment (alexpovel) [inline src/main.rs:1395]:
```suggestion
        /// The query can be given inline or as a path to a file containing a query.
```

**Comment by bheylin** [comment]
>     * I don't know how to report this better. On "file not found" error, we could check "does this LOOK like a file path", but I think that's fraught with peril... maybe a very simple, best-effort regex roughly like `/?([a-zA-Z0-9]+/)+`, and then if that matches `warn!()` the user.
>       Alternatively, is it possible the queries ALWAYS have to start with `(`? If yes, that could also be a give-away.

Yea I would avoid trying to guess if the string "looks" like a file.
I like the suggestion of having a leading symbol. That leading symbol actually ties in with
#158. I was going to suggest replacing `--query` and `--prepared` with a single `--query` arg
the uses a prefix symbol when using a prepared query. Something like:

```bash
# run literal query
srgn --query '((macro_invocation macro: (identifier) @_SRGN_IGNORE_name) @macro)'
# run `test-mod` prepared query
srgn --query '#test-mod'
# run contents of `./query/file.scm`
srgn --query ./query/file.scm`
```

File: README.md [hunk 11]
```diff
@@ -1191,6 +1191,17 @@ which can be caught as:
...
 ```
 
+##### Custom queries from file
+
+Typing out tree-sitter queries at the CLI can be unwieldy. To mitigate this you can read queries from [file](docs/query/python_cond_query.scm).
+
+Below we use the same Python file from the previous section with an invocation of
+
+```bash
+cat cond.py | srgn --python-query 'docs/python_cond_query.scm' --fail-any # will fail
```
[the remark this answers is not in the document]
Comment (bheylin) [reply README.md:1201]: Yea that's the output I expect. It's the same query as the section above.

File: src/main.rs (hunk 9 above)
Comment (bheylin) [reply src/main.rs:1479]: I don't think it's possible to know at this point what the contents of the `String` given to us from `clap` is.
We only know that it's a file path when we try open it later on.

Of course if we decided to use a prefix symbol to signal the intent of the `String` contents, then we could parse the `String` into this type of enum.

File: src/main.rs [hunk 12]
```diff
@@ -274,7 +274,7 @@ fn handle_actions_on_stdin(
     info!("Will use stdin to stdout.");
     let mut source = String::new();
     io::stdin().lock().read_to_string(&mut source)?;
-    let mut destination = String::with_capacity(source.len());
+    let mut destination = String::new();
```
[the remark this answers is not in the document]
Comment (bheylin) [reply src/main.rs:277]: Hmmm I'm not sure what happened here?
I'll fix it one way or the other.

File: src/main.rs [hunk 13]
```diff
@@ -209,8 +209,8 @@ fn main() -> Result<()> {
                 &language_scopers,
                 &actions,
                 search_mode,
-                options.threads.map_or_else(
```
[the remark this answers is not in the document]
Comment (bheylin) [reply src/main.rs]: Fixed in 66db56a6e4efd7b60e38d63647d6a8676a0f38d1

File: src/main.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (bheylin) [reply src/main.rs]: Fixed in 66db56a6e4efd7b60e38d63647d6a8676a0f38d1

File: src/main.rs [hunk 14]
```diff
@@ -1321,24 +1322,51 @@ mod cli {
...
+            .read(true)
+            .open(query_or_path.as_str())
+        {
+            Ok(mut file) => {
+                let mut s = String::new();
+                file.read_to_string(&mut s)?;
+                Ok(QuerySource::from(s))
+            }
+            Err(err) => match err.kind() {
+                io::ErrorKind::NotFound => Ok(QuerySource::from(query_or_path.0)),
```
[the remark this answers is not in the document]
Comment (bheylin) [reply src/main.rs]: Fixed in 3114b8db9adfbc1c51aa49f79d7926be106fcfc9

File: README.md [hunk 15]
```diff
@@ -1560,8 +1575,9 @@ Language scopes:
           - declaration:     Declaration
           - call-expression: Call expression
 
-      --c-query <TREE-SITTER-QUERY>
-          Scope C code using a custom tree-sitter query.
+      --c-query <TREE-SITTER-QUERY-OR-FILENAME>
+          Scope C code using a custom tree-sitter query. The query can be given inline or
+          if as a path to a file containing a query.
```
Comment (alexpovel) [inline README.md:1580]: Sorry, the `if` needs removing in a bunch of places now, i.e.

```suggestion
          as a path to a file containing a query.
```

**Review by alexpovel — APPROVED** [review]
CI still has minor problems, but happy to merge otherwise.

I might look into improving the error message/query-file-path-disambiguation more in the future.

**Comment by bheylin** [comment]
> CI still has minor problems, but happy to merge otherwise.
> 
> I might look into improving the error message/query-file-path-disambiguation more in the future.

All good, I think this PR is a nice contained change. The disambiguation issue is definitely for another PR.

**Comment by alexpovel** [comment]
I asked about what tree-sitter queries can potentially look like at: https://github.com/tree-sitter/tree-sitter/discussions/3862#discussioncomment-11120542 . The reply makes me think disambiguation isn't going to be straightforward. Even if we get it correct now, updates to tree-sitter might break us.

Currently, I'm leaning more towards a dedicated flag, like `--<LANG>-query-file`, again. That would get use niceties from `clap` as well, as it can natively handle `PathBuf`. We'd be more verbose but 100% safe going forward, leaving the normal tree-sitter `--<LANG>-query` untouched. 

Apologies for beating around the bush here, and changing my mind. I think I originally suggested overloading the existing option.

What do you think about merging this PR as a `chore`, then I can base the `feat` off it? I'm happy to take that work, having changed my mind on you and all that (and still not being sure what to do 😄 )

**Comment by bheylin** [comment]
> I asked about what tree-sitter queries can potentially look like at: [tree-sitter/tree-sitter#3862 (comment)](https://github.com/tree-sitter/tree-sitter/discussions/3862#discussioncomment-11120542) . The reply makes me think disambiguation isn't going to be straightforward. Even if we get it correct now, updates to tree-sitter might break us.
>
> Currently, I'm leaning more towards a dedicated flag, like `--<LANG>-query-file`, again. That would get use niceties from `clap` as well, as it can natively handle `PathBuf`. We'd be more verbose but 100% safe going forward, leaving the normal tree-sitter `--<LANG>-query` untouched.
>

Yea I agree, it's best to be as explicit as possible.
 
> Apologies for beating around the bush here, and changing my mind. I think I originally suggested overloading the existing option.
> 

No worries, that's how exploring an implementation works.

> What do you think about merging this PR as a `chore`, then I can base the `feat` off it? I'm happy to take that work, having changed my mind on you and all that (and still not being sure what to do 😄 )

I can also make the change, it's no problem at all.

**Comment by alexpovel** [comment]
Alright! Thanks for your work again and your patience.

Your change is merged now. The first of us to get to it can now build on top of that to implement something like `--<lang>-query-file` (just an example). I want to focus my time on bug fixes and other features for the immediate time being though.
